File: Assets/Code/Outfit7/Raven/TrackComponents/Events/Common/RavenBookmarkEvent.cc
#ifdef STARLITE
#include "RavenBookmarkEvent.h"
#include "RavenBookmarkEvent.cs"

#include <Utility/RavenLog.h>
#include <RavenSequence.h>

using namespace Starlite::Raven::Internal;

namespace Starlite {
    namespace Raven {
        RavenBookmarkEvent::RavenBookmarkEvent() {
        }

        bool RavenBookmarkEvent::IsBarrier() const {
            return false;
        }

        ERavenEventType RavenBookmarkEvent::GetEventType() const {
            return ERavenEventType::Bookmark;
        }

        int RavenBookmarkEvent::GetEndFrame() const {
            return m_StartFrame;
        }

        int RavenBookmarkEvent::GetLastFrame() const {
            return m_StartFrame;
        }

        RavenBookmarkEvent::ERavenBookmarkType RavenBookmarkEvent::GetBookmarkType() const {
            return m_BookmarkType;
        }

        const RavenAnimationDataComponentBase* RavenBookmarkEvent::GetAnimationDataEditorOnly() {
            return nullptr;
        }

        RavenPropertyComponent* RavenBookmarkEvent::GetPropertyComponent() const {
            return nullptr;
        }

        const String& RavenBookmarkEvent::GetBookmarkName() const {
            return m_BookmarkName;
        }

        void RavenBookmarkEvent::Initialize(Ptr<RavenSequence> sequence) {
            RavenEvent::Initialize(sequence);
            m_Sequence = sequence;
        }

        void RavenBookmarkEvent::OnEnter(int frame) {
#if RAVEN_DEBUG
            pRavenLog->InfoT(RavenSequence::Tag.GetCString(), "%s OnEnter %d", this, frame);
#endif
            m_ConditionsMet = ConditionsMet();

            if (m_ConditionsMet) {
                switch (m_BookmarkType) {
                case ERavenBookmarkType::Pause:
                    m_Sequence->Pause();
                    break;

                case ERavenBookmarkType::Stop:
                    m_Sequence->Stop();
                    break;

                case ERavenBookmarkType::Loop:
                    String prevBookmark = m_Sequence->FindFirstBookmarkBefore(m_StartFrame);
                    if (prevBookmark != String()) {
                        m_Sequence->GoToBookmark(prevBookmark);
                    } else {
                        m_Sequence->JumpToFrame(0);
                    }
                    break;
                }
            }
        }

        void RavenBookmarkEvent::RecalculateFpsChange(double durationFactor) {
            m_StartFrame = (int)(m_StartFrame * durationFactor);
        }

        void RavenBookmarkEvent::SetBookmarkName(String value) {
            m_BookmarkName = value;
        }

        void RavenBookmarkEvent::SetBookmarkType(RavenBookmarkEvent::ERavenBookmarkType value) {
            m_BookmarkType = value;
        }

        void RavenBookmarkEvent::SetEndFrame(int frame) {
            SetStartFrame(frame);
        }

        void RavenBookmarkEvent::SetLastFrame(int frame) {
            SetStartFrame(frame);
        }

        void RavenBookmarkEvent::SetTargetEditor(Ptr<RavenSequence> sequence, Ref<SceneObject>& target) {
        }

        int RavenBookmarkEvent::Comparer(const RavenBookmarkEvent* event1, const RavenBookmarkEvent* event2) {
            return RavenEvent::Comparer(event1, event2);
        }
    } // namespace Raven
} // namespace Starlite

#endif

File: Assets/Code/Outfit7/Raven/TrackComponents/Events/Common/RavenBookmarkEvent.cs
//{ "engine_register_objects":["Starlite__Raven__RavenBookmarkEvent"] }
#if !__cplusplus
#pragma warning disable
namespace Starlite {

    namespace Raven {
    
        // RavenBookmarkEvent
        [Starlite.ObjectClassAttribute("Starlite::Raven::RavenBookmarkEvent")]
        public sealed partial class RavenBookmarkEvent : global::Starlite.Raven.RavenEvent {
            // ERavenBookmarkType
            public enum ERavenBookmarkType { 
            	Ignore = 0, 
            	Pause = 1, 
            	Stop = 2, 
            	Loop = 3, 
            }
            
        	// Properties
        	[UnityEngine.SerializeField] [Starlite.ObjectPropertyAttribute(Starlite.SourceCCProperty.PropertyAttribute.Preload)] private Starlite.Raven.RavenBookmarkEvent.ERavenBookmarkType m_BookmarkType = (Starlite.Raven.RavenBookmarkEvent.ERavenBookmarkType) 0;
        	[UnityEngine.SerializeField] [Starlite.ObjectPropertyAttribute(Starlite.SourceCCProperty.PropertyAttribute.Preload)] private string m_BookmarkName = "Le Bookmark";
        }
        
    }
    
}
#pragma warning restore
#else
// RavenBookmarkEvent
namespace Starlite { namespace Raven { 
SIMPLEMENT_CLASS_WITH_PROPERTIES_IN_CLASS_SCOPED(Starlite::Raven::RavenBookmarkEvent, RavenBookmarkEvent, RavenBookmarkEvent, Starlite::Raven::RavenEvent, Starlite::SceneObjectComponent, "")
SIMPLEMENT_PROPERTY_UNIVERSAL(RavenBookmarkEvent, m_BookmarkType, m_BookmarkType, kInt, kUnknown, -1, "")
SIMPLEMENT_PROPERTY_UNIVERSAL(RavenBookmarkEvent, m_BookmarkName, m_BookmarkName, kString, kUnknown, -1, "")
SIMPLEMENT_CLASS_END(RavenBookmarkEvent)
} } 
SIMPLEMENT_CLASS_SCOPE(Starlite__Raven__RavenBookmarkEvent, Starlite::Raven, RavenBookmarkEvent)
#endif
